From fab001e7034ae14022659797a4e90f2f4fb63020 Mon Sep 17 00:00:00 2001
From: Daniel Stone <daniels@collabora.com>
Date: Thu, 9 Feb 2017 13:58:35 +0000
Subject: [PATCH 21/94] compositor-drm: Avoid connector_allocator for hotplugs

Rather than using connector_allocator to determine whether an output is
newly connected or not, use a list walk across all outputs instead.

Signed-off-by: Daniel Stone <daniels@collabora.com>
Reviewed-by: Quentin Glidic <sardemff7+git@sardemff7.net>
Reported-by: Peter Senna Tschudin <peter.senna@collabora.com>
---
 src/compositor-drm.c | 47 ++++++++++++++++++++++++++++++-----------------
 1 file changed, 30 insertions(+), 17 deletions(-)

diff --git a/src/compositor-drm.c b/src/compositor-drm.c
index 984e4ad..c8d85c7 100644
--- a/src/compositor-drm.c
+++ b/src/compositor-drm.c
@@ -257,6 +257,19 @@ drm_output_find_by_crtc(struct drm_backend *b, uint32_t crtc_id)
 	return NULL;
 }
 
+static struct drm_output *
+drm_output_find_by_connector(struct drm_backend *b, uint32_t connector_id)
+{
+	struct drm_output *output;
+
+	wl_list_for_each(output, &b->compositor->output_list, base.link) {
+		if (output->connector_id == connector_id)
+			return output;
+	}
+
+	return NULL;
+}
+
 static void
 drm_fb_destroy_callback(struct gbm_bo *bo, void *data)
 {
@@ -2673,6 +2686,9 @@ update_outputs(struct drm_backend *b, struct udev_device *drm_device)
 	/* collect new connects */
 	for (i = 0; i < resources->count_connectors; i++) {
 		int connector_id = resources->connectors[i];
+		struct weston_output *last =
+			container_of(b->compositor->output_list.prev,
+				     struct weston_output, link);
 
 		connector = drmModeGetConnector(b->drm.fd, connector_id);
 		if (connector == NULL)
@@ -2685,24 +2701,21 @@ update_outputs(struct drm_backend *b, struct udev_device *drm_device)
 
 		connected |= (1 << connector_id);
 
-		if (!(b->connector_allocator & (1 << connector_id))) {
-			struct weston_output *last =
-				container_of(b->compositor->output_list.prev,
-					     struct weston_output, link);
-
-			/* XXX: not yet needed, we die with 0 outputs */
-			if (!wl_list_empty(&b->compositor->output_list))
-				x = last->x + last->width;
-			else
-				x = 0;
-			y = 0;
-			create_output_for_connector(b, resources,
-						    connector, x, y,
-						    drm_device);
-			weston_log("connector %d connected\n", connector_id);
-
+		if (drm_output_find_by_connector(b, connector_id)) {
+			drmModeFreeConnector(connector);
+			continue;
 		}
-		drmModeFreeConnector(connector);
+
+		/* XXX: not yet needed, we die with 0 outputs */
+		if (!wl_list_empty(&b->compositor->output_list))
+			x = last->x + last->width;
+		else
+			x = 0;
+		y = 0;
+		create_output_for_connector(b, resources,
+					    connector, x, y,
+					    drm_device);
+		weston_log("connector %d connected\n", connector_id);
 	}
 	drmModeFreeResources(resources);
 
-- 
1.9.1

